idx/memory/memory.go

```diff
@@ -918,15 +919,29 @@ func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]idx.Archive, error) {
 				}
 			}
 			if staleCount == len(n.Defs) {
-				log.Debug("memory-idx: series %s for orgId:%d is stale. pruning it.", n.Path, org)
 				//we need to delete this node.
-				defs := m.delete(org, n, true)
-				statMetricsActive.Dec()
-				pruned = append(pruned, defs...)
+				toPrune = append(toPrune, n.Path)
 			}
 		}
-		m.Unlock()
+		m.RUnlock()
+
+		for _, path := range toPrune {
+			m.Lock()
+			n, ok := tree.Items[path]
+			if !ok {
+				m.Unlock()
+				log.Debug("memory-idx: series %s for orgId:%d was identified for pruning but cannot be found.", path, org)
+				continue
+			}
+
+			log.Debug("memory-idx: series %s for orgId:%d is stale. pruning it.", n.Path, org)
+			defs := m.delete(org, n, true)
+			statMetricsActive.Dec()
```

Comment:
Doesn't the `if !n.Leaf() {` check on line 912 mean that there should never be children?  We could add that same check here in case of a race condition where a child is added to a series that's about to be pruned, but that seems like a highly unlikely scenario.

Reply:
> Doesn't the if !n.Leaf() { check on line 912 mean that there should never be children? 

the only thing that branch does is assure that nodes without metricdefs (data corresponding to that path) are not subject to pruning.
nodes with metricdefs (leaf nodes) are subject to pruning.  I think the confusion here is because 
in our implementation, a path can be both a leaf and have children.
e.g. `foo.bar` can have data/metricdef (making it a leaf), while `foo.bar.baz` also exists (making foo.bar a branch)
In graphite this doesn't work, but some of our customers wanted this behavior, and it seemed fairly easy to support, so that's why we do.

the `m.delete` call is recursive: it deletes the path, and any sub-paths ([I think i actually just discovered a related bug here](https://github.com/grafana/metrictank/issues/714#issuecomment-353354166)), so what we should do is track the actual amount of paths that got deleted by the delete call, or I think better: the pruning process shouldn't recursively delete, but rather only delete leaves (and clean up stale branch nodes that no longer have children)
But this warrants a new ticket, so i opened https://github.com/grafana/metrictank/issues/797